sky/clouds/service_catalog/data_fetchers/fetch_gcp_metadata.py

```diff
@@ -0,0 +1,177 @@
+    return df
+
+
+def assert_zones_are_valid(zones: List[str]) -> None:
+    # Double-check that all the regions and zones are valid.
+    gcp_vm_zones_df = pd.read_html(GCP_VM_ZONES_URL)[0]
+    gcp_zones = gcp_vm_zones_df['Zones'].unique()
+    for zone in zones:
+        assert zone in gcp_zones
+
+
+def get_vm_df() -> Tuple[pd.DataFrame, List[str]]:
+    vm_df = pd.read_csv(VM_CSV_URL, delimiter=';')
+    # Drop deprecated VMs.
+    vm_df = vm_df[vm_df['DEPRECATED'].isna()]
+    # Drop unused columns.
+    vm_df = vm_df[['NAME', 'CPUS', 'MEMORY_GB', 'ZONE']]
+    # Rename the columns.
+    vm_df = vm_df.rename(
+        columns={
+            'NAME': 'InstanceType',
+            'CPUS': 'vCPUs',
+            'MEMORY_GB': 'MemoryGiB',
+            'ZONE': 'AvailabilityZone',
+        })
+    # Add the Region column.
+    vm_df['Region'] = vm_df['AvailabilityZone'].apply(lambda zone: zone[:-2])
+
+    # Drop invalid zones.
+    vm_df = drop_invalid_zones(vm_df)
+
+    # Convert vCPUs and MemoryGiB to float.
+    vm_df['vCPUs'] = vm_df['vCPUs'].astype(float)
+    vm_df['MemoryGiB'] = vm_df['MemoryGiB'].astype(float)
+
+    # Double-check that all the regions and zones are valid.
+    assert_zones_are_valid(vm_df['AvailabilityZone'].unique())
+
+    # Reorder the columns.
+    vm_df = vm_df[[
+        'InstanceType',
+        'vCPUs',
+        'MemoryGiB',
+        'Region',
+        'AvailabilityZone',
+    ]]
+    vm_df = vm_df.reset_index(drop=True)
+
+    # Get the list of zones that support 16xA100.
+    a2_megagpu_16g_df = vm_df[vm_df['InstanceType'] == 'a2-megagpu-16g']
+    a2_megagpu_16g_zones = a2_megagpu_16g_df['AvailabilityZone'].unique()
+    return vm_df, a2_megagpu_16g_zones
+
+
+def get_gpu_df(a2_megagpu_16g_zones: List[str]) -> pd.DataFrame:
+    gpu_df = pd.read_csv(GPU_CSV_URL, delimiter=';')
+    # Rename the columns.
+    gpu_df = gpu_df.rename(columns={
+        'NAME': 'AcceleratorName',
+        'ZONE': 'AvailabilityZone',
+    })
+    # Add the Region column.
+    gpu_df['Region'] = gpu_df['AvailabilityZone'].apply(lambda zone: zone[:-2])
+
+    # Drop VWS GPUs.
+    gpu_df = gpu_df[~gpu_df['AcceleratorName'].str.contains('vws')]
+    # Rename the GPUs.
+    gpu_df['AcceleratorName'] = gpu_df['AcceleratorName'].apply(
+        lambda name: GPU_NAMES[name])
+
+    # Drop invalid zones.
+    gpu_df = drop_invalid_zones(gpu_df)
+
+    # Add the AcceleratorCount column.
+    gpu_df['AcceleratorCount'] = gpu_df['AcceleratorName'].apply(
+        lambda acc_name: GPU_TYPES_TO_COUNTS[acc_name])
+    gpu_df = gpu_df.explode('AcceleratorCount', ignore_index=True)
+    gpu_df['AcceleratorCount'] = gpu_df['AcceleratorCount'].astype(int)
+
+    # Remove 16xA100 machines from zones that do not support them.
+    gpu_df = gpu_df[~((gpu_df['AcceleratorName'] == 'A100') &
+                      (gpu_df['AcceleratorCount'] == 16) &
+                      (~gpu_df['AvailabilityZone'].isin(a2_megagpu_16g_zones)))]
```

Comment:
does this mean that `GPU_CSV_URL` contain incorrect information so we have to manually remove them?

Reply:
The webpage provides all the zones that support any A2 machine. Here, the code first adds all A100 types (i.e., 1xA100, 2xA100, 4xA100, 8xA100, 16xA100) to those zones, and then removes 16xA100 from the zones that do not support it.